lib/srv/db/postgres/engine.go

```diff
@@ -0,0 +1,454 @@
+	}
+	return &pgproto3.ErrorResponse{
+		Severity: pgErr.Severity,
+		Code:     pgErr.Code,
+		Message:  pgErr.Message,
+		Detail:   pgErr.Detail,
+	}
+}
+
+// HandleConnection processes the connection from Postgres proxy coming
+// over reverse tunnel.
+//
+// It handles all necessary startup actions, authorization and acts as a
+// middleman between the proxy and the database intercepting and interpreting
+// all messages i.e. doing protocol parsing.
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *session.Context, clientConn net.Conn) (err error) {
+	client := pgproto3.NewBackend(pgproto3.NewChunkReader(clientConn), clientConn)
+	defer func() {
+		if err != nil {
+			if err := client.Send(toErrorResponse(err)); err != nil {
+				e.Log.WithError(err).Error("Failed to send error to client.")
+			}
+		}
+	}()
+	// The proxy is supposed to pass a startup message it received from
+	// the psql client over to us, so wait for it and extract database
+	// and username from it.
+	err = e.handleStartup(client, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Now we know which database/username the user is connecting to, so
+	// perform an authorization check.
+	err = e.checkAccess(sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// This is where we connect to the actual Postgres database.
+	server, hijackedConn, err := e.connect(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Upon successful connect, indicate to the Postgres client that startup
+	// has been completed and it can start sending queries.
+	err = e.makeClientReady(client, hijackedConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// At this point Postgres client should be ready to start sending
+	// messages: this is where psql prompt appears on the other side.
+	err = e.OnSessionStart(*sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		err := e.OnSessionEnd(*sessionCtx)
+		if err != nil {
+			e.Log.Error("Failed to emit audit event.")
+		}
+	}()
+	// Reconstruct pgconn.PgConn from hijacked connection for easier access
+	// to its utility methods (such as Close).
+	serverConn, err := pgconn.Construct(hijackedConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		err = serverConn.Close(ctx)
+		if err != nil {
+			e.Log.Error("Failed to close connection.")
+		}
+	}()
+	// Now launch the message exchange relaying all intercepted messages b/w
+	// the client (psql or other Postgres client) and the server (database).
+	clientErrCh := make(chan error, 1)
+	serverErrCh := make(chan error, 1)
+	go e.receiveFromClient(client, server, clientErrCh, sessionCtx)
+	go e.receiveFromServer(server, client, serverConn, serverErrCh, sessionCtx)
+	select {
+	case err := <-clientErrCh:
+		e.Log.WithError(err).Debug("Client done.")
+	case err := <-serverErrCh:
+		e.Log.WithError(err).Debug("Server done.")
+	case <-ctx.Done():
+		e.Log.Debug("Context canceled.")
+	}
+	return nil
```

Comment:
Who's responsible for closing `clientConn`?

Reply:
Right now all client connections are closed by the dispatchers (i.e. `db.ProxyServer.HandleConnection()` and `db.Server.HandleConnection()`), not specific proxy/engine implementations (of which there's just 1 so far).

This particular one is closed by `db.Server`, right where it first comes in over reverse tunnel: https://github.com/gravitational/teleport/pull/5005/files#diff-e7e5d004e7a2806bb6526efc22104150695206e576edd43e858cd621e89a1051R298.